rfc/rfc-65/rfc-65.md

```diff
@@ -0,0 +1,163 @@
+   decide which partitions are expired.
+2. User need to set different policies for different partitions. For example, the hudi table is partitioned
+   by `user_id`. For partition(user_id='1'), we set the policy that this partition will expire 7 days after its last
+   modified time, while for
+   partition(user_id='2') we can set the policy that it will expire 30 days after its last modified time.
+3. It's possible that users have multi-fields partitioning, and they don't want set policy for all partition. So we need
+   to support partition regex when defining
+   partition TTL policies. For example, for a hudi table partitioned by (user_id, ts), we can first set a default
+   policy(which policy value contains `*`)
+   that for all partitions matched (user_id=*/) will expire in 7 days, and explicitly set a policy that for partition
+   matched user_id=3 will expire in 30 days. Explicit policy will override default polices, which means default policies
+   only takes effects on partitions that do not match any explicit ttl policies.
+4. We may need to support record level ttl policy in the future. So partition TTL policy should be an implementation of
+   HoodieTTLPolicy.
+
+So we have the TTL policy definition like this (may change when implementing):
+
+```java
+public class HoodiePartitionTTLPolicy implements HoodieTTLPolicy {
+  public enum TTLPolicy {
+    // We supports keep by last modified time at the first version
+    KEEP_BY_TIME
+  }
+
+  // Partition spec for which the policy takes effect, could be a regex or a static partition value
+  private String partitionSpec;
+
+  private TTLPolicy policy;
+
+  private long policyValue;
+}
+```
+
+### User Interface for TTL policy Definition
+
+Users can config partition TTL management policies through SparkSQL Call Command and through table config directly.
+Assume that the user has a hudi table partitioned by two fields(user_id, ts), he can config partition TTL policies as
+follows.
+
+```sparksql
+-- Set policy for all user_id using partition regex, which keeps the data for 30 days.
+call add_ttl_policy(table => 'test', partitionSpec => 'user_id=*/', policy => 'KEEP_BY_TIME', policyValue => '30');
+
+--For partition user_id=3/, keep the data for 7 day.
+call add_ttl_policy(table => 'test', partitionSpec => 'user_id=3/', policy => 'KEEP_BY_TIME', policyValue => '7');
+
+-- Show all the TTL policies including default and explicit policies
+call show_ttl_policies(table => 'test');
+user_id=*/	KEEP_BY_TIME	30
+user_id=3/	KEEP_BY_TIME	7
+```
+
+### Storage for TTL policy
+
+We need persistent partition ttt policies in hudi table config and users should interact with hudi only by
+setting/getting ttl policies. Hudi then will takes care of doing the ttl management in an async table service.
+
+The partition TTL policies will be stored in `hoodie.properties` and they are defined as follows. Explicit policies are
```

Comment:
Is there any compatibility problem in storing the `hoodie.partition.ttl.policies` in `hoodie.properties`?

Reply:
Thanks for reminding me about this!  We can follow the common process of adding a new field to hoodie.properties. Will take it into consideration when implementing

Reply:
@stream2000, IMO, we could introduce `hoodie.ttl.properties` to store the `hoodie.partition.ttl.policies`, which have no compatibility problem here.